app/server/MergedServer.ts

```diff
@@ -0,0 +1,234 @@
+ *
+ */
+
+import {FlexServer, FlexServerOptions} from 'app/server/lib/FlexServer';
+import log from 'app/server/lib/log';
+import {getGlobalConfig} from "app/server/lib/globalConfig";
+
+// Allowed server types. We'll start one or a combination based on the value of GRIST_SERVERS
+// environment variable.
+export type ServerType = "home" | "docs" | "static" | "app";
+const allServerTypes: ServerType[] = ["home", "docs", "static", "app"];
+
+// Parse a comma-separate list of server types into an array, with validation.
+export function parseServerTypes(serverTypes: string|undefined): ServerType[] {
+  // Split and filter out empty strings (including the one we get when splitting "").
+  const types = (serverTypes || "").trim().split(',').filter(part => Boolean(part));
+
+  // Check that parts is non-empty and only contains valid options.
+  if (!types.length) {
+    throw new Error(`No server types; should be a comma-separated list of ${allServerTypes.join(", ")}`);
+  }
+  for (const t of types) {
+    if (!allServerTypes.includes(t as ServerType)) {
+      throw new Error(`Invalid server type '${t}'; should be in ${allServerTypes.join(", ")}`);
+    }
+  }
+  return types as ServerType[];
+}
+
+function checkUserContentPort(): number | null {
+  // Check whether a port is explicitly set for user content.
+  if (process.env.GRIST_UNTRUSTED_PORT) {
+    return parseInt(process.env.GRIST_UNTRUSTED_PORT, 10);
+  }
+  // Checks whether to serve user content on same domain but on different port
+  if (process.env.APP_UNTRUSTED_URL && process.env.APP_HOME_URL) {
+    const homeUrl = new URL(process.env.APP_HOME_URL);
+    const pluginUrl = new URL(process.env.APP_UNTRUSTED_URL);
+    // If the hostname of both home and plugin url are the same,
+    // but the ports are different
+    if (homeUrl.hostname === pluginUrl.hostname &&
+        homeUrl.port !== pluginUrl.port) {
+      const port = parseInt(pluginUrl.port || '80', 10);
+      return port;
+    }
+  }
+  return null;
+}
+
+interface ServerOptions extends FlexServerOptions {
+  // If set, messages logged to console (default: false)
+  // (but if options are not given at all in call to main, logToConsole is set to true)
+  logToConsole?: boolean;
+
+  // If set, documents saved to external storage such as s3 (default is to check environment variables,
+  // which get set in various ways in dev/test entry points)
+  externalStorage?: boolean;
+}
+
+export class MergedServer {
+
+  public static async create(port: number, serverTypes: ServerType[], options: ServerOptions = {}) {
+    options.settings ??= await getGlobalConfig();
+    const ms = new MergedServer(port, serverTypes, options);
+    // We need to know early on whether we will be serving plugins or not.
+    if (ms.hasComponent("home")) {
+      const userPort = checkUserContentPort();
+      ms.flexServer.setServesPlugins(userPort !== undefined);
+    } else {
+      ms.flexServer.setServesPlugins(false);
+    }
+
```

Comment:
Code to handle `options.loginSystem` is removed because login system has been moved to `ICreate`. FlexServer will now deal with the login system.